internal/unmarshal.go

```diff
@@ -38,135 +39,92 @@ func convertNumber[T Numbers](anyNumber any) (T, error) {
+func isTypeDataclass(typ reflect.Type) bool {
+	if typ.Kind() == reflect.Struct {
+		for i := 0; i < typ.NumField(); i++ {
+			fieldMeta := typ.Field(i)
+			if fieldMeta.Tag.Get("dataclass_field") == "true" {
+				return true
+			}
 		}
-		typedSlice[i] = castRes
 	}
-	return typedSlice, nil
+	return false
 }
 
-func convertIfNumber(value any, kind reflect.Kind) (any, error) {
-	// TODO: Figure out if we could possibly
-	// do the equivalent by creating a new
-	// reflect.Value with reflect.New
-	// and reflect.Value.Set.
-	var err error
-	switch kind {
-	case reflect.Int8:
-		value, err = convertNumber[int8](value)
-	case reflect.Int16:
-		value, err = convertNumber[int16](value)
-	case reflect.Int32:
-		value, err = convertNumber[int32](value)
-	case reflect.Int64:
-		value, err = convertNumber[int64](value)
-	case reflect.Uint8:
-		value, err = convertNumber[uint8](value)
-	case reflect.Uint16:
-		value, err = convertNumber[uint16](value)
-	case reflect.Uint32:
-		value, err = convertNumber[uint32](value)
-	case reflect.Uint64:
-		value, err = convertNumber[uint64](value)
-	case reflect.Float32:
-		value, err = convertNumber[float32](value)
-	case reflect.Float64:
-		value, err = convertNumber[float64](value)
-	}
-
-	return value, err
+func IsDataclass(field reflect.Value) bool {
+	return isTypeDataclass(field.Type())
 }
 
-func convertSlice(sliceElemKind reflect.Kind, value any) (any, error) {
-	anySlice := value.([]any)
-	switch sliceElemKind {
-	case reflect.Int8:
-		return convertSliceyNumbers[int8](anySlice)
-	case reflect.Int16:
-		return convertSliceyNumbers[int16](anySlice)
-	case reflect.Int32:
-		return convertSliceyNumbers[int32](anySlice)
-	case reflect.Int64:
-		return convertSliceyNumbers[int64](anySlice)
-	case reflect.Uint8:
-		return convertSliceyNumbers[uint8](anySlice)
-	case reflect.Uint16:
-		return convertSliceyNumbers[uint16](anySlice)
-	case reflect.Uint32:
-		return convertSliceyNumbers[uint32](anySlice)
-	case reflect.Uint64:
-		return convertSliceyNumbers[uint64](anySlice)
-	case reflect.Float32:
-		return convertSliceyNumbers[float32](anySlice)
-	case reflect.Float64:
-		return convertSliceyNumbers[float64](anySlice)
-	case reflect.String:
-		return convertSliceyNonNumbers[string](anySlice)
-	case reflect.Bool:
-		return convertSliceyNonNumbers[bool](anySlice)
-	default:
-		return nil, fmt.Errorf("unsupported slice type '%s' when converting slice", sliceElemKind)
+func GetValueFromArrowArray(a arrow.Array, idx int) (any, error) {
+	if a.IsNull(idx) {
+		return nil, nil
 	}
-}
-
-func getPointerToCopied(elemType reflect.Type, value any) reflect.Value {
-	copied := reflect.New(reflect.TypeOf(value))
-	copied.Elem().Set(reflect.ValueOf(value))
-	castedPointer := reflect.NewAt(elemType, copied.UnsafePointer())
-	return castedPointer
-}
-
-func GetValueFromArrowArray(arr arrow.Array, idx int) (any, error) {
-	switch castArr := arr.(type) {
+	switch arr := a.(type) {
+	case *array.LargeList:
+		newSlice := make([]any, 0)
```

Comment:
nit: as a perf opt, could we preallocate the newSlice to have capacity `arr.Offsets()[idx+1] - arr.Offsets()[idx]`

Reply:
thanks!